idx/memory/tag_query.go

```diff
@@ -0,0 +1,180 @@
+package memory
+
+import (
+	"errors"
+	"fmt"
+	"regexp"
+	"strings"
+
+	"github.com/raintank/metrictank/idx"
+)
+
+var (
+	expressionRe    = regexp.MustCompile("^([^!=~]+)([!=~]{1,3})(.*)$")
+	errUnknownId    = errors.New("id does not exist")
+	errInvalidTag   = errors.New("found an invalid tag in the index")
+	errInvalidQuery = errors.New("invalid query")
+)
+
+type TagQuery struct {
+	selects []func(TagIndex) (map[string]struct{}, error)
+	filters []func(def *idx.Archive) (bool, error)
+}
+
+func NewTagQuery(expressions []string) (*TagQuery, error) {
+	if len(expressions) == 0 {
+		return nil, errInvalidQuery
+	}
+
+	query := &TagQuery{}
+
+	for _, expr := range expressions {
+		if !expressionRe.MatchString(expr) {
+			return nil, errInvalidQuery
+		}
+
+		exprSplits := expressionRe.FindAllStringSubmatch(expr, -1)
+		if len(exprSplits[0]) != 4 {
+			return nil, errInvalidQuery
+		}
+
+		key := exprSplits[0][1]
+		operator := exprSplits[0][2]
+		value := exprSplits[0][3]
+
+		// always anchor all regular expressions at the beginning
+		if operator[len(operator)-1] == byte('~') && len(value) > 0 && value[len(value)-1] != byte('^') {
+			value = fmt.Sprintf("^%s", value)
+		}
+
+		// special case of empty value
+		if len(value) == 0 {
+		}
+
+		switch operator {
+		case "=":
+			query.selects = append(query.selects, func(idx TagIndex) (map[string]struct{}, error) {
+				var values map[string]map[string]struct{}
+				var ok bool
+				if values, ok = idx[key]; !ok {
+					return make(map[string]struct{}), nil
+				}
+				if ids, ok := values[value]; ok {
+					return ids, nil
+				}
+				return make(map[string]struct{}), nil
+			})
+		case "=~":
+			query.selects = append(query.selects, func(idx TagIndex) (map[string]struct{}, error) {
+				res := make(map[string]struct{})
+				var values map[string]map[string]struct{}
+				var ok bool
+				if values, ok = idx[key]; !ok {
+					return res, nil
+				}
+
+				re, err := regexp.Compile(value)
+				if err != nil {
+					return nil, err
+				}
+
+				for v, ids := range values {
```

Comment:
same question, it seems like this could just be `for v, ids := range idx[key] {`